Mirror Refinements 1: added Fidelity types to schema module

wpank opened the pull request:
This commit addresses (#998) and (#996) and corresponds to
1.) in @wchargin's suggested implementation plan.

The following adds a `Fidelity` type to Mirror.

`Fidelity` is either:
    - `FAITHFUL`
    - `UNFAITHFUL`

An `UNFAITHFUL` type has an `actualTypenames` property that lists
all the types of objects that can be returned when the field is
queried.

`fidelity` thus is added to Objects that are prone to inconsistent
typenames. These are `NodeFieldType` and `NestedFieldType`.

If any schema contains a Node or Nested Node  with a field that
contains an `UNFAITHFUL` `fidelity` type, a `Not Yet Implemented`
Error is thrown.

Test Plan:

A test has been added to the constructor to check that an error
is thrown given a schema that contains unfaithful fields.

Added a function for creating an unfaithful github schema,
`buildGithubSchemaUnfaithful`.

Added tests to `schema.test.js` to check validity of Unfaithful
Fidelity and duplicate `actualTypenames`

wchargin commented:
I’ll try to review this tonight.

wchargin commented on src/graphql/mirror.js:
```diff
@@ -1133,6 +1133,11 @@ export class Mirror {
             case "NODE":
+              if (field.fidelity.type === "UNFAITHFUL") {
+                throw new Error(
+                  "Handling unfaithful fields is not yet implemented"
```
Are these two assertions reachable? At a glance, it looks like we should
never get here due to the constructor validation. If that’s correct,
then please replace the error text with something that clearly indicates
that this is an invariant violation (as in other errors in this file),
and mark the enclosing conditionals with `// istanbul ignore if` to keep
the coverage report in sync with the assertions.

wchargin commented on src/graphql/mirror.js:
```diff
@@ -1133,6 +1133,11 @@ export class Mirror {
             case "NODE":
+              if (field.fidelity.type === "UNFAITHFUL") {
+                throw new Error(
+                  "Handling unfaithful fields is not yet implemented"
```
Are these two assertions reachable? At a glance, it looks like we should
never get here due to the constructor validation. If that’s correct,
then please replace the error text with something that clearly indicates
that this is an invariant violation (as in other errors in this file),
and mark the enclosing conditionals with `// istanbul ignore if` to keep
the coverage report in sync with the assertions.

wpank replied:
This is not reachable and is taken out in a later commit. Updated the error text :+1:

wchargin commented on src/graphql/schema.test.js:
```diff
@@ -152,6 +152,24 @@ describe("graphql/schema", () => {
         'field "O"/"foo" has invalid type "Color" of kind "ENUM"'
       );
     });
+    it("disallows objects with node fields of unfaithful type", () => {
```
This test name is wrong. Node fields of unfaithful type are and should
be allowed. What’s disallowed is unfaithful fidelities whose lists of
actual typenames include a type that does not exist.

wchargin commented on src/graphql/schema.test.js:
```diff
@@ -152,6 +152,24 @@ describe("graphql/schema", () => {
         'field "O"/"foo" has invalid type "Color" of kind "ENUM"'
       );
     });
+    it("disallows objects with node fields of unfaithful type", () => {
```
This test name is wrong. Node fields of unfaithful type are and should
be allowed. What’s disallowed is unfaithful fidelities whose lists of
actual typenames include a type that does not exist.

wpank replied:
:+1: changed

wchargin commented on src/graphql/schema.test.js:
```diff
@@ -152,6 +152,24 @@ describe("graphql/schema", () => {
+        O: s.object({id: s.id(), foo: s.node("Color", s.unfaithful(["Tree"]))}),
+      };
+      expect(() => Schema.schema(types2)).toThrowError(
+        'field "O"/"foo" has invalid actualTypename Tree in its unfaithful fidelity'
```
Please add a test that constructs a valid schema with unfaithful fields.
How about just modifying the existing `buildGithubSchema`?

wpank replied:
:+1: done

wchargin commented on src/graphql/mirror.test.js:
```diff
@@ -267,6 +339,44 @@ describe("graphql/mirror", () => {
         expect(() => new Mirror(db, schema0)).toThrow("invalid field name");
         expect(() => new Mirror(db, buildGithubSchema())).not.toThrow();
       });
+      it("rejects a schema with unfaithful fidelities", () => {
```
dWhenever you have `schema0`, `schema1`, … and they’re independent of
each other, that’s a sign that you probably want to split things up! :-)
Thankfully, you can buy more test cases for pretty cheap these days.

wchargin replied:
Whenever you have `schema0`, `schema1`, … and they’re independent of
each other, that’s a sign that you probably want to split things up! :-)
Thankfully, you can buy more test cases for pretty cheap these days.

wpank replied:
:+1:  done

wchargin commented on src/graphql/generateFlowTypes.js:
```diff
@@ -48,12 +48,18 @@ export default function generateFlowTypes(
     }
   }
   function formatNodeField(field: NodeFieldType): string {
+    if (field.fidelity.type === "UNFAITHFUL") {
```
Did you test the changes to `generateFlowTypes`?

I’m somewhat sympathetic to only lightly testing changes that consist of
adding error cases that are planned to be removed in a subsequent
commit, but these seem simple enough to test that you might as well do
it. Personally, I wouldn’t feel comfortable checking in functional
changes to a file that is already well-tested without updating the tests
at all.

wpank replied:
:+1: done

wchargin commented:
(Looks like some of my comments were duplicated on submission—sorry
about that. As mentioned on Discord, my internet is trying very hard to
keep up at the moment. Will try to clean these up tomorrow; feel free to
disregard duplicates, of course.)

wchargin commented:
(Looks like some of my comments were duplicated on submission—sorry
about that. As mentioned on Discord, my internet is trying very hard to
keep up at the moment. Will try to clean these up tomorrow; feel free to
disregard duplicates, of course.)

wpank commented:
@wchargin thanks for the detailed response!

> What were your thoughts about connections when designing this code?
Do you plan to add them in later? Did you just want to cut scope?

I initially thought including connections was to address an unfaithful `Actor` in a connection (thus also considering that connection as unfaithful). My main concern was to address the blacklist and blocked repositories. 

I'm now looking at [your comment in #996 about IssueComments](https://github.com/sourcecred/sourcecred/issues/996#issuecomment-435495537). I don't think this affects Sourcecred at the moment, since the schema doesn't include an `issue` connection, nor `pullRequest` connection. Correct me if I'm wrong, but you can only get an `IssueComment` from the `Issue` or `Pull Request` that it comes from right? If that's true then in those cases you should already have the appropriate typename for the node that `IssueComment` belongs to , since the `Issue` or `Pull Request` gotten from a repository _should_ be consistent (they aren't a connection to any other object). 

I can make another issue to address handling that, but I think that sort of thing would be best to contact Github.

wpank commented on src/graphql/schema.js:
```diff
@@ -272,8 +307,26 @@ export function nullable(elementType: Typename): PrimitiveTypeAnnotation {
+  },
+  fidelity?: Fidelity = faithful()
+): NestedFieldType {
+  return {type: "NESTED", eggs: {...eggs}, fidelity};
```
Can you elaborate on this a little bit? Do you mean to add a test: `if (a.fidelity !== "FAITHFUL") throw new Error()` to `is deep-equal to but deep-distinct from its input`?

wchargin replied:
When the user passes a value into the “constructor-like” functions in
this module, we always shallow-copy that value. This is why we create a
new object in `nested` rather than using the original `eggs`, and why in
`schema` we create new objects for each of the field types.

Just as `nested` shallow-copies its eggs, so too should it shallow-copy
its fidelity. Same for `node`. You probably want to define a helper
function `_cloneFidelity` to handle the `switch`-`case`ing. You need not
test this helper function, though it should be tested indirectly.

Does that make sense?

wpank commented on src/graphql/schema.test.js:
```diff
@@ -152,6 +152,24 @@ describe("graphql/schema", () => {
+
+      expect(() => {
+        s.node("Color", s.unfaithful(["Color", "Color"]));
+      }).toThrowError("duplicate unfaithful typename Color");
```
>that the map is injective modulo permutation of inputs

Can you elaborate on this?

wchargin replied:
Elaborated here:
<https://github.com/sourcecred/sourcecred/pull/1095#pullrequestreview-206666987>

(GitHub doesn’t let you reply to an inline comment on a commit that has
been force-pushed if you have an pending review. This behavior is
actually kind of my fault, because it’s their somewhat hacky fix to a
nasty bug that I reported. But whatever.)

wpank commented on src/graphql/schema.js:
```diff
@@ -272,8 +307,26 @@ export function nullable(elementType: Typename): PrimitiveTypeAnnotation {
+  const actualTypenames: {|[Typename]: true|} = ({}: any);
+  for (const t of typenames) {
+    if (actualTypenames[t] === true) {
+      throw new Error(`duplicate unfaithful typename ${t}`);
```
:+1:  done

wpank commented on src/graphql/schema.js:
```diff
@@ -115,6 +123,28 @@ export function schema(types: {[Typename]: NodeType}): Schema {
+                }
+                return;
+              default:
+                // istanbul ignore next: unreachable per Flow
```
:+1:  done

wpank commented on src/graphql/schema.js:
```diff
@@ -115,6 +123,28 @@ export function schema(types: {[Typename]: NodeType}): Schema {
+                  const fidelityType = types[possibleTypename];
+                  if (fidelityType == null) {
+                    throw new Error(
+                      `${self} has invalid actualTypename ${possibleTypename} in its unfaithful fidelity`
```
:+1:  done

wpank commented on src/graphql/mirror.test.js:
```diff
@@ -105,6 +105,78 @@ describe("graphql/mirror", () => {
     }
     return s.schema(types);
   }
+  function buildGithubSchemaUnfaithful(): Schema.Schema {
```
:+1: done.

wpank commented:
Updated the commit and code to address all the above minus the two questions ^.  Also I changed my git `user.name` config to wpank, but for some reason the log file still shows the author as just `will`. Is that coming from elsewhere?
